WalletWasabi/WabiSabi/Client/CoinJoinClient.cs

```diff
@@ -0,0 +1,202 @@
+		private CredentialPool VsizeCredentialPool { get; } = new();
+		private ClientRound Round { get; }
+		public IArenaRequestHandler ArenaRequestHandler { get; }
+		public Kitchen Kitchen { get; }
+		public KeyManager Keymanager { get; }
+		private SecureRandom SecureRandom { get; }
+		private CancellationTokenSource DisposeCts { get; } = new();
+		private Coin[] Coins { get; set; }
+		private Random Random { get; } = new();
+
+		public CoinJoinClient(
+			ClientRound round,
+			IArenaRequestHandler arenaRequestHandler,
+			IEnumerable<Coin> coins,
+			Kitchen kitchen,
+			KeyManager keymanager)
+		{
+			Round = round;
+			ArenaRequestHandler = arenaRequestHandler;
+			Kitchen = kitchen;
+			Keymanager = keymanager;
+			SecureRandom = new SecureRandom();
+			Coins = coins.ToArray();
+		}
+
+		protected override async Task ExecuteAsync(CancellationToken stoppingToken)
+		{
+			try
+			{
+				// Register coins.
+				AliceClient[] aliceClients = await RegisterCoinsAsync(stoppingToken).ConfigureAwait(false);
+
+				// Confirm coins.
+				await ConfirmConnectionsAsync(aliceClients, stoppingToken).ConfigureAwait(false);
+
+				// Planning
+				var outputs = DecomposeAmounts(stoppingToken);
+
+				// Output registration.
+				await ReissueAndRegisterOutputsAsync(outputs, stoppingToken).ConfigureAwait(false);
+
+				Transaction? unsignedCoinJoinTransaction = null; // TODO: Get it from somewhere.
+
+				// Send signature.
+				await SignTransactionAsync(aliceClients, unsignedCoinJoinTransaction).ConfigureAwait(false);
+			}
+			catch (Exception ex)
+			{
+				// The game is over for this round, no fallback mechanism. In the next round we will create another CoinJoinClient and try again.
+			}
+		}
+
+		public async Task StartMixingCoinsAsync()
+		{
+			await StartAsync(DisposeCts.Token).ConfigureAwait(false);
+		}
+
+		private async Task<AliceClient[]> RegisterCoinsAsync(CancellationToken stoppingToken)
+		{
+			var aliceArenaClient = new ArenaClient(
+				Round.AmountCredentialIssuerParameters,
+				Round.VsizeCredentialIssuerParameters,
+				AmountCredentialPool,
+				VsizeCredentialPool,
+				ArenaRequestHandler,
+				SecureRandom);
+
+			List<AliceClient> aliceClients = new();
+
+			foreach (var coin in Coins)
+			{
+				var secret = Keymanager.GetSecrets(Kitchen.SaltSoup(), coin.ScriptPubKey.WitHash.ScriptPubKey).First().PrivateKey.GetBitcoinSecret(Keymanager.GetNetwork());
+				aliceClients.Add(await AliceClient.CreateNewAsync(aliceArenaClient, coin, secret, Round.Id, Round.FeeRate).ConfigureAwait(false));
+				await Task.Delay(Random.Next(0, 1000), stoppingToken).ConfigureAwait(false);
+			}
+
+			return aliceClients.ToArray();
+		}
+
+		private async Task ConfirmConnectionsAsync(IEnumerable<AliceClient> aliceClients, CancellationToken stoppingToken)
+		{
+			foreach (var alice in aliceClients)
+			{
+				await alice.ConfirmConnectionAsync(TimeSpan.FromMilliseconds(Random.Next(1000, 5000)), stoppingToken).ConfigureAwait(false);
```

Comment:
I would prefer this in two lines just to use a variable as an intermediate value holder that help me to understand what the random timespan is for. Take into account that **all** the requests already have a randomized delay (plus the one added by the tor network).

Also, I would prefer using the `InsecureRandom.Random` because that is a global random instance and I remember reading something about an issue with new `Random`, I don't remember why exactly but reusing the same `Random` instance provides you a better random numbers.

```suggestion
				var confirmationInterval =TimeSpan.FromMilliseconds(Random.Next(1000, 5000)
				await alice.ConfirmConnectionAsync(confirmationInterval), stoppingToken).ConfigureAwait(false);
```

Reply:
I was wrong. The delay shouldn't be here, it has to be in the `WabiSabiApiClient`. We should merge that first perhaps.
Then we can simplify this as follow:

```c#
var confirmationRequests = aliceClients.Select(client => 
    client.ConfirmConnectionAsync(stoppingToken);
await Task.WhenAll(confirmationRequests, stoppingToken).ConfigureAwait(false);
```

Reply:
You mean `WabiSabiHttpApiClient.cs`?